File: api/src/main/java/org/zenika/api/controller/PokemonsController.java
package org.zenika.api.controller;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.zenika.core.Battle;
import org.zenika.core.Pokemon;
import org.zenika.api.service.PokemonService;

import java.util.List;

@RestController
@RequestMapping("/pokemons")
public class PokemonsController {

    private PokemonService pokemonService;

    public PokemonsController(PokemonService pokemonService) {
        this.pokemonService = pokemonService;
    }

    @GetMapping
    public List<Pokemon> getAllPokemons() {
        return pokemonService.getAllPokemon();
    }

    @GetMapping("/{idOrName}")
    public ResponseEntity<Pokemon> getPokemonById(@PathVariable String idOrName) {
        Pokemon pokemon = pokemonService.getPokemonByIdOrName(idOrName);
        if (pokemon == null) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(pokemon);
    }

    @GetMapping("/types")
    public List<String> getPokemonTypes() {
        return pokemonService.getAllPokemonTypes();
    }

    @GetMapping("/{idPokemon1}/vs/{idPokemon2}")
    public ResponseEntity<Battle> fight(@PathVariable String idPokemon1, @PathVariable String idPokemon2) {
        Battle battle = pokemonService.fight(idPokemon1, idPokemon2);
        if (battle == null) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(battle);
    }

    @GetMapping("/battle/{id}")
    public ResponseEntity<Battle> getBattle(@PathVariable String id) {
        Battle battle = pokemonService.getBattle(id);
        if (battle == null) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(battle);
    }
}

File: api/src/main/kotlin/org/zenika/api/controller/PokemonsController.kt
package org.zenika.api.controller

import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.zenika.api.service.PokemonService
import org.zenika.core.Battle
import org.zenika.core.Pokemon

@RestController
@RequestMapping("/pokemons")
class PokemonsController(val pokemonService: PokemonService) {

    private fun <T> T?.wrap(): ResponseEntity<T> =
            this?.let { ResponseEntity.ok(it) } ?: ResponseEntity.notFound().build()


    @GetMapping
    fun getAllPokemons(): List<Pokemon> = pokemonService.getAllPokemon()

    @GetMapping("/{idOrName}")
    fun getPokemonById(@PathVariable idOrName: String): ResponseEntity<Pokemon> =
            pokemonService.getPokemonByIdOrName(idOrName).wrap()

    @GetMapping("/types")
    fun getPokemonTypes() : List<String> = pokemonService.getAllPokemonTypes()

    @GetMapping("/{idPokemon1}/vs/{idPokemon2}")
    fun fight(@PathVariable idPokemon1: String, @PathVariable idPokemon2: String): ResponseEntity<Battle> =
            pokemonService.fight(idPokemon1, idPokemon2).wrap()

    @GetMapping("/battle/{id}")
    fun getBattle(@PathVariable id: String): ResponseEntity<Battle> = pokemonService.getBattle(id).wrap()
}
